torch/autograd/functional.py

```diff
@@ -0,0 +1,714 @@
+            a tuple of Tensor or a Tensor.
+        inputs (tuple of Tensor or Tensor): inputs to the function.
+        v (tuple of Tensor or Tensor): The vector that will multiply the Jacobian. Must be the
+            same size as the input of ``func``. This argument is optional when
+            ``func``'s input contains a single element and it will be set as a Tensor
+            containing a single ``1``.
+        create_graph (bool, optional): If ``True``, both the output and result will be
+            computed in a differentiable way. Note that when strict if ``False``, the result can not
+            require gradients or be disconnected from the inputs.
+            Defaults to ``False``.
+        strict (bool, optional): If ``True``, an error will be raised when we detect we detect that some quantities
+            are independent of the element we want to differentiate it with. If ``False``, such quantity
+            will only be set to ``0`` which is the expected mathematical value.
+            Defaults to ``False``.
+
+    Returns:
+        output (tuple of Tensor or Tensor): output of ``func(inputs)``
+        result (tuple of Tensor or Tensor): result of the dot product with the same shape
+            as the output.
+
+    Example::
+
+        >>> def exp_reducer(x):
+        ...   return x.exp().sum(dim=1)
+        >>> inputs = torch.rand(4, 4)
+        >>> v = torch.ones(4, 4)
+        >>> jvp(exp_reducer, inputs, v)
+        (tensor([6.3090, 4.6742, 7.9114, 8.2106]),
+         tensor([6.3090, 4.6742, 7.9114, 8.2106]))
+
+        >>> jvp(exp_reducer, inputs, v, create_graph=True)
+        (tensor([6.3090, 4.6742, 7.9114, 8.2106], grad_fn=<SumBackward1>),
+         tensor([6.3090, 4.6742, 7.9114, 8.2106], grad_fn=<SqueezeBackward1>))
+
+        >>> def adder(x, y):
+        ...   return 2 * x + 3 * y
+        >>> inputs = (torch.rand(2), torch.rand(2))
+        >>> v = (torch.ones(2), torch.ones(2))
+        >>> jvp(adder, inputs, v)
+        (tensor([2.2399, 2.5005]),
+         tensor([5., 5.]))
+    """
+
+    tuple_inputs, inputs = _as_tuple(inputs, "inputs", "jvp")
+    inputs = _grad_preprocess(inputs, create_graph=create_graph, need_graph=True)
+
+    if v is not None:
+        tuple_v, v = _as_tuple(v, "v", "jvp")
+        v = _grad_preprocess(v, create_graph=create_graph, need_graph=False)
+        _validate_v(v, inputs)
+    else:
+        if len(inputs) != 1 or inputs[0].nelement() != 1:
+            raise RuntimeError("The vector v can only be None if the input to the user-provided function "
+                               "is a single Tensor with a single element.")
+
+    outputs = func(*inputs)
+    tuple_outputs, outputs = _as_tuple(outputs, "outputs of the user-provided function", "jvp")
+    _check_requires_grad(outputs, "outputs", strict=strict)
+    grad_outputs = tuple(torch.zeros_like(out, requires_grad=True) for out in outputs)
```

Comment:
Does it matter what the grad_outputs is for the double backward trick?

Reply:
No, it's a linear function, so the backward is independent on the value (if there is no nan or inf). I do want to set it to a valid value though to avoid NaNs that could appear with `torch.empty()`. Do you think I should add a comment?

Reply:
A comment to a resource about the double backward trick would probably be helpful for anyone looking at this code in the future